### PR TITLE
Migrate Java client emitter

### PR DESCRIPTION
**Pull request by srnagar**

### CODE REVIEW

File: pnpm-lock.yaml
Comment (timotheeguerin) [inline pnpm-lock.yaml:1]: probably shouldn't have change to this file

File: packages/http-client-java/emitter/.prettierignore
Comment (timotheeguerin) [inline packages/http-client-java/emitter/.prettierignore:1]: I don't think we need this file, the one at the root of the repo will already contain everything we need an dprettier doesn't support nested ones

File: packages/http-client-java/emitter/tsconfig.json [hunk 1]
```diff
@@ -0,0 +1,27 @@
+{
+  "compilerOptions": {
```
Comment (timotheeguerin) [inline packages/http-client-java/emitter/tsconfig.json:2]: I think might be able to extend the typespec base one like csharp did

File: packages/http-client-java/package.json [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+    "url": "https://github.com/Microsoft/typespec/issues"
+  },
+  "license": "MIT",
+  "type": "module",
+  "main": "emitter/dist/src/emitter.js",
+  "exports": {
+    ".": "./emitter/dist/src/emitter.js",
+    "./testing": "./dist/src/testing/index.js"
+  },
+  "tspMain": "emitter/dist/src/emitter.js",
```
Comment (timotheeguerin) [inline packages/http-client-java/package.json:25]: Don't need `tspMain`

File: packages/http-client-java/package.json [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+  "engines": {
+    "node": ">=14.0.0"
```
Comment (timotheeguerin) [inline packages/http-client-java/package.json:27]:
```suggestion
    "node": ">=18.0.0"
```

File: packages/http-client-java/package.json [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+  },
+  "scripts": {
+    "clean": "rimraf .emitter/dist ./temp",
+    "build:emitter": "tsc -p ./emitter/tsconfig.build.json",
```
Comment (timotheeguerin) [inline packages/http-client-java/package.json:31]: Recommend you also have a `build` one 

csharp for example
```
"build": "npm run build:emitter && npm run build:generator && npm run extract-api",
```

File: packages/http-client-java/package.json [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+    "emitter/dist/**",
+    "!emitter/dist/test/**",
+    "generator/http-client-generator/target/emitter.jar"
+  ],
+  "peerDependencies": {
+    "@azure-tools/typespec-azure-core": ">=0.43.0 <1.0.0",
+    "@azure-tools/typespec-azure-resource-manager": ">=0.43.0 <1.0.0",
+    "@azure-tools/typespec-autorest": ">=0.43.0 <1.0.0",
+    "@azure-tools/typespec-azure-rulesets": ">=0.43.0 <1.0.0",
+    "@azure-tools/typespec-client-generator-core": ">=0.43.2 <1.0.0",
```
Comment (timotheeguerin) [inline packages/http-client-java/package.json:49]: are all those deps needed here

File: packages/http-client-java/package.json [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+    "@typespec/http": ">=0.57.0 <1.0.0",
+    "@typespec/openapi": ">=0.57.0 <1.0.0",
+    "@typespec/rest": ">=0.57.0 <1.0.0",
+    "@typespec/versioning": ">=0.57.0 <1.0.0",
+    "@typespec/xml": ">=0.57.0 <1.0.0"
+  },
+  "dependencies": {
+    "@autorest/codemodel": "~4.20.0",
+    "js-yaml": "~4.1.0",
+    "lodash": "~4.17.21"
```
Comment (timotheeguerin) [inline packages/http-client-java/package.json:60]: Try not to use lodash

File: packages/http-client-java/package.json [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+    "@azure-tools/typespec-client-generator-core": ">=0.43.2 <1.0.0",
+    "@typespec/compiler": ">=0.57.0 <1.0.0",
+    "@typespec/http": ">=0.57.0 <1.0.0",
+    "@typespec/openapi": ">=0.57.0 <1.0.0",
+    "@typespec/rest": ">=0.57.0 <1.0.0",
+    "@typespec/versioning": ">=0.57.0 <1.0.0",
+    "@typespec/xml": ">=0.57.0 <1.0.0"
+  },
+  "dependencies": {
+    "@autorest/codemodel": "~4.20.0",
```
Comment (timotheeguerin) [inline packages/http-client-java/package.json:58]: Is autorest codemodel used?

File: packages/http-client-java/package.json [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+    "@typescript-eslint/parser": "~7.8.0",
+    "@typespec/compiler": "0.57.0",
+    "@typespec/http": "0.57.0",
+    "@typespec/openapi": "0.57.0",
+    "@typespec/rest": "0.57.0",
+    "@typespec/versioning": "0.57.0",
+    "c8": "~9.1.0",
+    "eslint": "~8.57.0",
+    "eslint-plugin-deprecation": "~2.0.0",
+    "mocha": "~10.4.0",
```
Comment (timotheeguerin) [inline packages/http-client-java/package.json:83]: Would recommend you move to vitest like csharp, it should be pretty simple

File: packages/http-client-java/test/package.json
Comment (timotheeguerin) [inline packages/http-client-java/test/package.json:1]: Not sure this extra package is very necessary? can't this be included in the base one

File: packages/http-client-java/package.json [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+    "@azure-tools/typespec-azure-resource-manager": "0.43.0",
+    "@azure-tools/typespec-autorest": "0.43.0",
+    "@azure-tools/typespec-azure-rulesets": "0.43.0",
+    "@azure-tools/typespec-client-generator-core": "0.43.0",
+    "@azure-tools/codegen": "~2.10.0",
+    "@types/js-yaml": "~4.0.9",
+    "@types/lodash": "~4.17.1",
+    "@types/mocha": "~10.0.6",
+    "@types/node": "~20.12.10",
+    "@typescript-eslint/eslint-plugin": "~7.8.0",
```
Comment (timotheeguerin) [inline packages/http-client-java/package.json:73]: Remove all the eslint packages, the typespec repo will already enforce that

**Review by timotheeguerin — COMMENTED** [review]
A few comments mostly on consitency across all our packages to try to use as much common tools as possible. Feel free to file some follow up issues for some of them and/or let me know if you ened some help

File: packages/http-client-java/package.json (hunk 6 above)
Comment (srnagar) [reply packages/http-client-java/package.json:60]: created an issue to remove this dependency - https://github.com/microsoft/typespec/issues/4051

File: packages/http-client-java/package.json [hunk 10]
```diff
@@ -58,6 +58,7 @@
     "@autorest/codemodel": "~4.20.0",
     "@azure-tools/ts-genapi": "^1.0.7",
     "@microsoft/api-extractor-model": "^7.28.2",
+    "@typespec/http-client-java": "file:",
```
Comment (timotheeguerin) [inline packages/http-client-java/package.json:61]: this feels weird to have, what is the reason?

Comment (srnagar) [reply packages/http-client-java/package.json]: hmm, this was auto-updated. Reverted the changes now.

File: packages/http-client-java/Setup.ps1 [hunk 11]
```diff
@@ -0,0 +1,15 @@
...
+if ($RebuildJar) {
+    mvn clean install -f generator/pom.xml -DskipTests
+    if ($LASTEXITCODE -ne 0) {
+        exit $LASTEXITCODE
+    }
+}
+
+# re-build http-client-java
+npm install 
+npm run build:generator && npm run build:emitter && npm pack
```
Comment (timotheeguerin) [inline packages/http-client-java/Setup.ps1:14]: is pack needed here? Feels like there is maybe some process simplification that could happen

File: packages/http-client-java/eng/scripts/Test-Packages.ps1
Comment (timotheeguerin) [inline packages/http-client-java/eng/scripts/Test-Packages.ps1:1]: Would be great if most those script could be simplified to be either just `package.json` scripts or called from one so one could just look at that list and do the common script commands
```bash
npm run build
npm run test
```

without having to learn a different set of scripts files for each emitter

File: packages/http-client-java/package.json [hunk 12]
```diff
@@ -0,0 +1,85 @@
...
+    "lib/*.tsp",
+    "emitter/dist/**",
+    "!emitter/dist/test/**",
+    "generator/http-client-generator/target/emitter.jar"
+  ],
+  "peerDependencies": {
+    "@azure-tools/typespec-autorest": ">=0.43.0 <1.0.0",
+    "@azure-tools/typespec-azure-core": ">=0.43.0 <1.0.0",
+    "@azure-tools/typespec-client-generator-core": ">=0.43.2 <1.0.0",
+    "@microsoft/api-extractor": "^7.40.3",
```
Comment (timotheeguerin) [inline packages/http-client-java/package.json:49]: I think api extractor should be a dev Dependency

File: packages/http-client-java/package.json [hunk 13]
```diff
@@ -0,0 +1,85 @@
...
+    "extract-api": "npx api-extractor run --local --verbose"
+  },
+  "files": [
+    "lib/*.tsp",
+    "emitter/dist/**",
+    "!emitter/dist/test/**",
+    "generator/http-client-generator/target/emitter.jar"
+  ],
+  "peerDependencies": {
+    "@azure-tools/typespec-autorest": ">=0.43.0 <1.0.0",
```
Comment (timotheeguerin) [inline packages/http-client-java/package.json:46]: is autorest needed?

File: packages/http-client-java/package.json [hunk 14]
```diff
@@ -0,0 +1,85 @@
...
+    "@azure-tools/typespec-azure-core": ">=0.43.0 <1.0.0",
+    "@azure-tools/typespec-client-generator-core": ">=0.43.2 <1.0.0",
+    "@microsoft/api-extractor": "^7.40.3",
+    "@typespec/compiler": ">=0.57.0 <1.0.0",
+    "@typespec/http": ">=0.57.0 <1.0.0",
+    "@typespec/openapi": ">=0.57.0 <1.0.0",
```
Comment (timotheeguerin) [inline packages/http-client-java/package.json:52]: Same with openapi

File: packages/http-client-java/package.json (hunk 5 above)
Comment (srnagar) [reply packages/http-client-java/package.json]: created an issue to evaluate the need for all these dependencies - https://github.com/microsoft/typespec/issues/4117

File: packages/http-client-java/package.json (hunk 13 above)
Comment (srnagar) [reply packages/http-client-java/package.json:45]: Will be investigated as part of this issue - https://github.com/microsoft/typespec/issues/4117

File: packages/http-client-java/package.json (hunk 14 above)
Comment (srnagar) [reply packages/http-client-java/package.json:51]: Will be investigated as part of this issue - https://github.com/microsoft/typespec/issues/4117

**Review by timotheeguerin — APPROVED** [review]

**Pull request merged**